<a href="https://colab.research.google.com/github/AbhayK007/NLP_Algorithms/blob/main/Spam_detection_NLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np


In [ ]:
data = pd.read_csv('/content/drive/MyDrive/IITB STUDY/Projects/SOC/3rd week NLP/SMSSpamCollection', sep ='\t', names=['label','mail'])

In [ ]:
data.head(5)

label                                               mail
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

Data Cleaning

In [ ]:
import nltk
import re
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')
from nltk.stem import WordNetLemmatizer
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer 

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [ ]:
ps= PorterStemmer()
wnl= WordNetLemmatizer()
corpus=[]

In [ ]:
for i in range(0,len(data)):
  review=re.sub('[^a-zA-z]',' ',data['mail'][i])
  review= review.lower()
  review= review.split()
  review= [wnl.lemmatize(word) for word in review if word not in set(stopwords.words('english'))]
  review= ' '.join(review)
  corpus.append(review)

Creating Bag of words

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer


In [ ]:
cv= TfidfVectorizer()

In [ ]:
X = cv.fit_transform(corpus).toarray()
print(X.shape)

(5572, 7103)


X has 7103 different words. but we dont need that many words. So we will select top frequent words

In [ ]:
CV=TfidfVectorizer(max_features=4000)

In [ ]:
X_4000=CV.fit_transform(corpus).toarray()
print(X_4000.shape)

(5572, 4000)


converting ham spam into binary

In [ ]:
y=pd.get_dummies(data['label'])
y.head()

ham  spam
0    1     0
1    1     0
2    0     1
3    1     0
4    1     0

In [ ]:
y=y.iloc[:,1].values
y

array([0, 0, 1, ..., 0, 0, 0], dtype=uint8)

In [ ]:
from sklearn.model_selection import train_test_split
xtrain,xtest,ytrain,ytest= train_test_split(X_4000,y,test_size=0.2)


training with multinomial naive bayes

In [ ]:
from sklearn.naive_bayes import MultinomialNB

In [ ]:
spam_model= MultinomialNB().fit(xtrain,ytrain)

In [ ]:
y_predict= spam_model.predict(xtest)

In [ ]:
from sklearn.metrics import accuracy_score

In [ ]:
accuracy = accuracy_score(ytest,y_predict)

In [ ]:
accuracy

0.9748878923766816